lib/smack/SmackRep.cpp

```diff
@@ -361,31 +474,44 @@ string SmackRep::getString(const llvm::Value* v) {
 const Expr* SmackRep::cast(const llvm::Instruction* I) {
   return cast(I->getOpcode(), I->getOperand(0), I->getType());
 }
-
 const Expr* SmackRep::cast(const llvm::ConstantExpr* CE) {
   return cast(CE->getOpcode(), CE->getOperand(0), CE->getType());
 }
 
+string SmackRep::opName(const string& operation, initializer_list<unsigned> operands) {
+  stringstream s;
+  s << operation;
+  for (initializer_list<unsigned>::const_iterator i = operands.begin(), e = operands.end(); i != e; ++i)
+    s << "." << int_type(*i);
+  return s.str();
+}
+
 const Expr* SmackRep::cast(unsigned opcode, const llvm::Value* v, const llvm::Type* t) {
   using namespace llvm;
   switch (opcode) {
-  case Instruction::Trunc:
-    assert(t->isIntegerTy() && "TODO: implement truncate for non-integer types.");
-    return isBool(t)
-      ? Expr::fn("$i2b",expr(v))
-      : Expr::fn("$trunc",expr(v),lit(t->getPrimitiveSizeInBits()));
-
-  case Instruction::ZExt:
-  case Instruction::SExt:
-    return isBool(v->getType()) ? b2p(v) : expr(v);
-
-  case Instruction::FPTrunc:
-  case Instruction::FPExt:
-  case Instruction::BitCast:
-    return expr(v);
-
-  default:
-    return Expr::fn(cast2fn(opcode), expr(v));
+    case Instruction::Trunc:
+      assert(t->isIntegerTy() && "TODO: implement truncate for non-integer types.");
+      return (isBool(t) && !SmackOptions::BitVectors)
```

Comment:
The `$trunc.iXX.iYY` expression should be generated in either case (bit-vectors or not). The `$i2b` expression should also be generated in either case, depending on the type of `t`.